File: src/main/scala/jagfx/synth/ToneSynthesizer.scala
package jagfx.synth

import jagfx.Constants
import jagfx.Constants.Int16
import jagfx.Constants.NoisePhaseDiv
import jagfx.Constants.PhaseMask
import jagfx.Constants.PhaseScale
import jagfx.model.*
import jagfx.utils.MathUtils.clipInt16

/** Synthesizes single `Voice` into audio samples using FM/AM modulation. */
object VoiceSynthesizer:
  /** Generates audio samples from `Voice` definition. */
  def synthesize(voice: Voice): AudioBuffer =
    val sampleCount = voice.duration * Constants.SampleRate / 1000
    if sampleCount <= 0 || voice.duration < 10 then return AudioBuffer.empty(0)

    val samplesPerStep = sampleCount.toDouble / voice.duration.toDouble
    val buffer = BufferPool.acquire(sampleCount)

    val state = initSynthState(voice, samplesPerStep)
    renderSamples(buffer, voice, state, sampleCount)

    applyGating(buffer, voice, sampleCount)
    applyEcho(buffer, voice, samplesPerStep, sampleCount)

    voice.filter.foreach { f =>
      FilterSynthesizer.apply(buffer, f, sampleCount)
    }

    clipInt16(buffer, sampleCount)

    val output = new Array[Int](sampleCount)
    System.arraycopy(buffer, 0, output, 0, sampleCount)
    BufferPool.release(buffer)
    AudioBuffer(output, Constants.SampleRate)

  // Types
  private case class SynthState(
      freqBaseEval: EnvelopeEvaluator,
      ampBaseEval: EnvelopeEvaluator,
      freqModRateEval: Option[EnvelopeEvaluator],
      freqModRangeEval: Option[EnvelopeEvaluator],
      ampModRateEval: Option[EnvelopeEvaluator],
      ampModRangeEval: Option[EnvelopeEvaluator],
      frequencyStart: Int,
      frequencyDuration: Int,
      amplitudeStart: Int,
      amplitudeDuration: Int,
      partialDelays: Array[Int],
      partialVolumes: Array[Int],
      partialSemitones: Array[Int],
      partialStarts: Array[Int]
  )

  private def initSynthState(voice: Voice, samplesPerStep: Double): SynthState =
    val freqBaseEval = EnvelopeEvaluator(voice.pitchEnvelope)
    val ampBaseEval = EnvelopeEvaluator(voice.volumeEnvelope)
    freqBaseEval.reset()
    ampBaseEval.reset()

    val (freqModRateEval, freqModRangeEval, vibratoLfoIncr, vibratoLfoBase) =
      initFrequencyModulation(voice, samplesPerStep)

    val (ampModRateEval, ampModRangeEval, amplitudeStart, amplitudeDuration) =
      initAmplitudeModulation(voice, samplesPerStep)

    val (delays, volumes, semitones, starts) =
      initPartials(voice, samplesPerStep)

    SynthState(
      freqBaseEval,
      ampBaseEval,
      freqModRateEval,
      freqModRangeEval,
      ampModRateEval,
      ampModRangeEval,
      vibratoLfoIncr,
      vibratoLfoBase,
      amplitudeStart,
      amplitudeDuration,
      delays,
      volumes,
      semitones,
      starts
    )

  private def initFrequencyModulation(
      voice: Voice,
      samplesPerStep: Double
  ): (Option[EnvelopeEvaluator], Option[EnvelopeEvaluator], Int, Int) =
    voice.vibratoRate match
      case Some(env) =>
        val rateEval = EnvelopeEvaluator(env)
        val rangeEval = voice.vibratoDepth.map(EnvelopeEvaluator(_))
        rateEval.reset()
        rangeEval.foreach(_.reset())
        val start = ((env.end - env.start) * PhaseScale / samplesPerStep).toInt
        val duration = (env.start * PhaseScale / samplesPerStep).toInt
        (Some(rateEval), rangeEval, start, duration)
      case None =>
        (None, None, 0, 0)

  private def initAmplitudeModulation(
      voice: Voice,
      samplesPerStep: Double
  ): (Option[EnvelopeEvaluator], Option[EnvelopeEvaluator], Int, Int) =
    voice.tremoloRate match
      case Some(env) =>
        val rateEval = EnvelopeEvaluator(env)
        val rangeEval = voice.tremoloDepth.map(EnvelopeEvaluator(_))
        rateEval.reset()
        rangeEval.foreach(_.reset())
        val start = ((env.end - env.start) * PhaseScale / samplesPerStep).toInt
        val duration = (env.start * PhaseScale / samplesPerStep).toInt
        (Some(rateEval), rangeEval, start, duration)
      case None =>
        (None, None, 0, 0)

  private def initPartials(
      voice: Voice,
      samplesPerStep: Double
  ): (Array[Int], Array[Int], Array[Int], Array[Int]) =
    val delays = new Array[Int](Constants.MaxPartials)
    val volumes = new Array[Int](Constants.MaxPartials)
    val semitones = new Array[Int](Constants.MaxPartials)
    val starts = new Array[Int](Constants.MaxPartials)

    for partial <- 0 until math.min(
        Constants.MaxPartials,
        voice.partials.length
      )
    do
      val height = voice.partials(partial)
      if height.volume.value != 0 then
        delays(partial) = (height.startDelay.value * samplesPerStep).toInt
        volumes(partial) = (height.volume.value << 14) / 100
        semitones(partial) =
          ((voice.pitchEnvelope.end - voice.pitchEnvelope.start) * PhaseScale *
            LookupTables.getPitchMultiplier(
              height.pitchOffset
            ) / samplesPerStep).toInt
        starts(partial) =
          (voice.pitchEnvelope.start * PhaseScale / samplesPerStep).toInt

    (delays, volumes, semitones, starts)

  private def renderSamples(
      buffer: Array[Int],
      voice: Voice,
      state: SynthState,
      sampleCount: Int
  ): Unit =
    val phases = new Array[Int](Constants.MaxPartials)
    var frequencyPhase = 0
    var amplitudePhase = 0

    for sample <- 0 until sampleCount do
      var frequency = state.freqBaseEval.evaluate(sampleCount)
      var amplitude = state.ampBaseEval.evaluate(sampleCount)

      val (newFreq, newFreqPhase) =
        applyVibrato(frequency, frequencyPhase, sampleCount, state, voice)
      frequency = newFreq
      frequencyPhase = newFreqPhase
      val (newAmp, newAmpPhase) =
        applyTremolo(amplitude, amplitudePhase, sampleCount, state, voice)
      amplitude = newAmp
      amplitudePhase = newAmpPhase

      renderPartials(
        buffer,
        voice,
        state,
        sample,
        sampleCount,
        frequency,
        amplitude,
        phases
      )

  private def applyVibrato(
      frequency: Int,
      phase: Int,
      sampleCount: Int,
      state: SynthState,
      voice: Voice
  ): (Int, Int) =
    (state.freqModRateEval, state.freqModRangeEval) match
      case (Some(rateEval), Some(rangeEval)) =>
        val rate = rateEval.evaluate(sampleCount)
        val range = rangeEval.evaluate(sampleCount)
        val mod =
          generateSample(range, phase, voice.vibratoRate.get.waveform) >> 1
        val nextPhase =
          phase + (rate * state.frequencyStart >> 16) + state.frequencyDuration
        (frequency + mod, nextPhase)
      case _ => (frequency, phase)

  private def applyTremolo(
      amplitude: Int,
      phase: Int,
      sampleCount: Int,
      state: SynthState,
      voice: Voice
  ): (Int, Int) =
    (state.ampModRateEval, state.ampModRangeEval) match
      case (Some(rateEval), Some(rangeEval)) =>
        val rate = rateEval.evaluate(sampleCount)
        val range = rangeEval.evaluate(sampleCount)
        val mod =
          generateSample(range, phase, voice.tremoloRate.get.waveform) >> 1
        val newAmp = amplitude * (mod + Constants.Int16.UnsignedMaxValue) >> 15
        val nextPhase =
          phase + (rate * state.amplitudeStart >> 16) + state.amplitudeDuration
        (newAmp, nextPhase)
      case _ => (amplitude, phase)

  private def renderPartials(
      buffer: Array[Int],
      voice: Voice,
      state: SynthState,
      sample: Int,
      sampleCount: Int,
      frequency: Int,
      amplitude: Int,
      phases: Array[Int]
  ): Unit =
    for partial <- 0 until math.min(
        Constants.MaxPartials,
        voice.partials.length
      )
    do
      if voice.partials(partial).volume.value != 0 then
        val position = sample + state.partialDelays(partial)
        if position >= 0 && position < sampleCount then
          buffer(position) += generateSample(
            amplitude * state.partialVolumes(partial) >> 15,
            phases(partial),
            voice.pitchEnvelope.waveform
          )
          phases(partial) += (frequency * state.partialSemitones(
            partial
          ) >> 16) +
            state.partialStarts(partial)

  private def generateSample(
      amplitude: Int,
      phase: Int,
      waveform: Waveform
  ): Int =
    waveform match
      case Waveform.Square =>
        if (phase & PhaseMask) < Int16.Quarter then amplitude else -amplitude
      case Waveform.Sine =>
        (LookupTables.sin(phase & PhaseMask) * amplitude) >> 14
      case Waveform.Saw =>
        (((phase & PhaseMask) * amplitude) >> 14) - amplitude
      case Waveform.Noise =>
        LookupTables.noise((phase / NoisePhaseDiv) & PhaseMask) * amplitude
      case Waveform.Off => 0

  private def applyGating(
      buffer: Array[Int],
      voice: Voice,
      sampleCount: Int
  ): Unit =
    (voice.gateSilence, voice.gateDuration) match
      case (Some(silence), Some(duration)) =>
        val silenceEval = EnvelopeEvaluator(silence)
        val durationEval = EnvelopeEvaluator(duration)
        silenceEval.reset()
        durationEval.reset()
        var counter = 0
        var muted = true
        for sample <- 0 until sampleCount do
          val stepOn = silenceEval.evaluate(sampleCount)
          val stepOff = durationEval.evaluate(sampleCount)
          val threshold =
            if muted then
              silence.start + ((silence.end - silence.start) * stepOn >> 8)
            else silence.start + ((silence.end - silence.start) * stepOff >> 8)
          counter += 256
          if counter >= threshold then
            counter = 0
            muted = !muted
          if muted then buffer(sample) = 0
      case _ => ()

  private def applyEcho(
      buffer: Array[Int],
      voice: Voice,
      samplesPerStep: Double,
      sampleCount: Int
  ): Unit =
    if voice.echoDelay > 0 && voice.echoMix > 0 then
      val start = (voice.echoDelay * samplesPerStep).toInt
      for sample <- start until sampleCount do
        buffer(sample) += buffer(sample - start) * voice.echoMix / 100

File: JagFX.Synthesis/ToneSynthesizer.cs
using JagFX.Domain;
using JagFX.Domain.Models;
using JagFX.Domain.Utilities;

namespace JagFX.Synthesis;

public static class ToneSynthesizer
{
    public static SampleBuffer Synthesize(Voice voice)
    {
        var sampleCount = (int)(voice.Duration * Constants.SampleRatePerMillisecond);
        if (sampleCount <= 0 || voice.Duration < 10)
        {
            return SampleBuffer.Empty(0);
        }

        var samplesPerStep = sampleCount / (double)voice.Duration;
        var buffer = SampleBufferPool.Acquire(sampleCount);

        var state = InitSynthState(voice, samplesPerStep);
        RenderSamples(buffer, voice, state, sampleCount);

        ApplyGating(buffer, voice, sampleCount);
        ApplyEcho(buffer, voice, samplesPerStep, sampleCount);

        if (voice.Filter != null)
        {
            FilterProcessor.Apply(buffer, voice.Filter, state.FilterEnvelopeEval, sampleCount);
        }

        MathUtils.ClipInt16(buffer, sampleCount);

        var output = new int[sampleCount];
        Array.Copy(buffer, 0, output, 0, sampleCount);

        SampleBufferPool.Release(buffer);
        return new SampleBuffer(output, Constants.SampleRate);
    }

    private static SynthesisState InitSynthState(Voice voice, double samplesPerStep)
    {
        var freqBaseEval = new EnvelopeEvaluator(voice.FrequencyEnvelope);
        var ampBaseEval = new EnvelopeEvaluator(voice.AmplitudeEnvelope);
        freqBaseEval.Reset();
        ampBaseEval.Reset();

        var (freqModRateEval, freqModRangeEval, vibratoStart, vibratoDuration) =
            InitFrequencyModulation(voice, samplesPerStep);

        var (ampModRateEval, ampModRangeEval, amplitudeStart, amplitudeDuration) =
            InitAmplitudeModulation(voice, samplesPerStep);

        var (delays, volumes, semitones, starts) =
            InitOscillators(voice, samplesPerStep);

        EnvelopeEvaluator? filterEnvelopeEval = null;
        if (voice.Filter != null && voice.Filter.Envelope != null)
        {
            filterEnvelopeEval = new EnvelopeEvaluator(voice.Filter.Envelope);
            filterEnvelopeEval.Reset();
        }

        return new SynthesisState
        {
            FreqBaseEval = freqBaseEval,
            AmpBaseEval = ampBaseEval,
            FreqModRateEval = freqModRateEval,
            FreqModRangeEval = freqModRangeEval,
            AmpModRateEval = ampModRateEval,
            AmpModRangeEval = ampModRangeEval,
            FrequencyStart = vibratoStart,
            FrequencyDuration = vibratoDuration,
            AmplitudeStart = amplitudeStart,
            AmplitudeDuration = amplitudeDuration,
            PartialDelays = delays,
            PartialVolumes = volumes,
            PartialSemitones = semitones,
            PartialStarts = starts,
            FilterEnvelopeEval = filterEnvelopeEval
        };
    }

    private static (EnvelopeEvaluator? rateEval, EnvelopeEvaluator? rangeEval, int start, int duration)
        InitFrequencyModulation(Voice voice, double samplesPerStep)
    {
        if (voice.PitchLfo != null)
        {
            var rateEval = new EnvelopeEvaluator(voice.PitchLfo.Rate);
            var rangeEval = new EnvelopeEvaluator(voice.PitchLfo.Depth);
            rateEval.Reset();
            rangeEval.Reset();

            var start = (int)((double)(voice.PitchLfo.Rate.End - voice.PitchLfo.Rate.Start) * Constants.PhaseScale / samplesPerStep);
            var duration = (int)(voice.PitchLfo.Rate.Start * Constants.PhaseScale / samplesPerStep);

            return (rateEval, rangeEval, start, duration);
        }

        return (null, null, 0, 0);
    }

    private static (EnvelopeEvaluator? rateEval, EnvelopeEvaluator? rangeEval, int start, int duration)
        InitAmplitudeModulation(Voice voice, double samplesPerStep)
    {
        if (voice.AmplitudeLfo != null)
        {
            var rateEval = new EnvelopeEvaluator(voice.AmplitudeLfo.Rate);
            var rangeEval = new EnvelopeEvaluator(voice.AmplitudeLfo.Depth);
            rateEval.Reset();
            rangeEval.Reset();

            var start = (int)((double)(voice.AmplitudeLfo.Rate.End - voice.AmplitudeLfo.Rate.Start) * Constants.PhaseScale / samplesPerStep);
            var duration = (int)(voice.AmplitudeLfo.Rate.Start * Constants.PhaseScale / samplesPerStep);

            return (rateEval, rangeEval, start, duration);
        }

        return (null, null, 0, 0);
    }

    private static (int[] delays, int[] volumes, int[] semitones, int[] starts)
        InitOscillators(Voice voice, double samplesPerStep)
    {
        var delays = new int[Constants.MaxOscillators];
        var volumes = new int[Constants.MaxOscillators];
        var semitones = new int[Constants.MaxOscillators];
        var starts = new int[Constants.MaxOscillators];

        var oscillatorCount = Math.Min(Constants.MaxOscillators, voice.Oscillators.Count);

        for (var oscillator = 0; oscillator < oscillatorCount; oscillator++)
        {
            var height = voice.Oscillators[oscillator];
            if (height.Amplitude.Value != 0)
            {
                delays[oscillator] = (int)(height.Delay.Value * samplesPerStep);
                volumes[oscillator] = (height.Amplitude.Value << 14) / 100;
                semitones[oscillator] = (int)(
                    (voice.FrequencyEnvelope.End - voice.FrequencyEnvelope.Start) * Constants.PhaseScale *
                    LookupTables.GetPitchMultiplier(height.PitchOffset) / samplesPerStep);
                starts[oscillator] = (int)(voice.FrequencyEnvelope.Start * Constants.PhaseScale / samplesPerStep);

                if (oscillator == 0)
                {
                    _ = LookupTables.GetPitchMultiplier(height.PitchOffset);
                }
                else if (oscillator == 1)
                {
                    _ = LookupTables.GetPitchMultiplier(height.PitchOffset);
                }
            }
        }

        return (delays, volumes, semitones, starts);
    }

    private static void RenderSamples(
        int[] buffer,
        Voice voice,
        SynthesisState state,
        int sampleCount)
    {
        var phases = new int[Constants.MaxOscillators];
        var frequencyPhase = 0;
        var amplitudePhase = 0;

        for (var sample = 0; sample < sampleCount; sample++)
        {
            var frequency = state.FreqBaseEval.Evaluate(sampleCount);
            var amplitude = state.AmpBaseEval.Evaluate(sampleCount);

            (frequency, frequencyPhase) = ApplyVibrato(frequency, frequencyPhase, sampleCount, state, voice);
            (amplitude, amplitudePhase) = ApplyTremolo(amplitude, amplitudePhase, sampleCount, state, voice);

            RenderOscillators(
                buffer,
                voice,
                state,
                sample,
                sampleCount,
                frequency,
                amplitude,
                phases);
        }
    }

    private static void RenderOscillators(
        int[] buffer,
        Voice voice,
        SynthesisState state,
        int sample,
        int sampleCount,
        int frequency,
        int amplitude,
        int[] phases)
    {
        var oscillatorCount = Math.Min(Constants.MaxOscillators, voice.Oscillators.Count);
        for (var oscillator = 0; oscillator < oscillatorCount; oscillator++)
        {
            if (voice.Oscillators[oscillator].Amplitude.Value != 0)
            {
                var position = sample + state.PartialDelays[oscillator];
                if (position >= 0 && position < sampleCount)
                {
                    var sampleValue = GenerateSample(
                        amplitude * state.PartialVolumes[oscillator] >> 15,
                        phases[oscillator],
                        voice.FrequencyEnvelope.Waveform);

                    buffer[position] += sampleValue;
                    phases[oscillator] += (frequency * state.PartialSemitones[oscillator] >> 16) +
                        state.PartialStarts[oscillator];
                }
            }
        }
    }

    private static (int frequency, int phase) ApplyVibrato(
        int frequency,
        int phase,
        int sampleCount,
        SynthesisState state,
        Voice voice)
    {
        if (state.FreqModRateEval != null && state.FreqModRangeEval != null)
        {
            var rate = state.FreqModRateEval.Evaluate(sampleCount);
            var range = state.FreqModRangeEval.Evaluate(sampleCount);
            var mod = GenerateSample(range, phase, voice.PitchLfo!.Rate.Waveform) >> 1;
            var nextPhase = phase + (rate * state.FrequencyStart >> 16) + state.FrequencyDuration;

            return (frequency + mod, nextPhase);
        }

        return (frequency, phase);
    }

    private static (int amplitude, int phase) ApplyTremolo(
        int amplitude,
        int phase,
        int sampleCount,
        SynthesisState state,
        Voice voice)
    {
        if (state.AmpModRateEval != null && state.AmpModRangeEval != null)
        {
            var rate = state.AmpModRateEval.Evaluate(sampleCount);
            var range = state.AmpModRangeEval.Evaluate(sampleCount);
            var mod = GenerateSample(range, phase, voice.AmplitudeLfo!.Rate.Waveform) >> 1;

            var newAmp = amplitude * (mod + Constants.FixedPoint.Offset) >> 15;
            var nextPhase = phase + (rate * state.AmplitudeStart >> 16) + state.AmplitudeDuration;
            return (newAmp, nextPhase);
        }

        return (amplitude, phase);
    }

    private static int GenerateSample(int amplitude, int phase, Waveform waveform)
    {
        return waveform switch
        {
            Waveform.Square => (phase & Constants.PhaseMask) < Constants.FixedPoint.Quarter ? amplitude : -amplitude,
            Waveform.Sine => (LookupTables.SinTable[phase & Constants.PhaseMask] * amplitude) >> 14,
            Waveform.Saw => (((phase & Constants.PhaseMask) * amplitude) >> 14) - amplitude,
            Waveform.Noise => LookupTables.NoiseTable[(phase / Constants.NoisePhaseDiv) & Constants.PhaseMask] * amplitude,
            Waveform.Off => 0,
            _ => 0
        };
    }

    private static void ApplyGating(int[] buffer, Voice voice, int sampleCount)
    {
        if (voice.GateSilence != null && voice.GateDuration != null)
        {
            var silenceEval = new EnvelopeEvaluator(voice.GateSilence);
            var durationEval = new EnvelopeEvaluator(voice.GateDuration);
            silenceEval.Reset();
            durationEval.Reset();

            var counter = 0;
            var muted = true;

            for (var sample = 0; sample < sampleCount; sample++)
            {
                var stepOn = silenceEval.Evaluate(sampleCount);
                var stepOff = durationEval.Evaluate(sampleCount);
                var threshold = muted
                    ? voice.GateSilence.Start + ((voice.GateSilence.End - voice.GateSilence.Start) * stepOn >> 8)
                    : voice.GateSilence.Start + ((voice.GateSilence.End - voice.GateSilence.Start) * stepOff >> 8);

                counter += 256;
                if (counter >= threshold)
                {
                    counter = 0;
                    muted = !muted;
                }

                if (muted)
                {
                    buffer[sample] = 0;
                }
            }
        }
    }

    private static void ApplyEcho(int[] buffer, Voice voice, double samplesPerStep, int sampleCount)
    {
        if (voice.FeedbackDelay.Delay > 0 && voice.FeedbackDelay.Mix > 0)
        {
            var start = (int)(voice.FeedbackDelay.Delay * samplesPerStep);
            for (var sample = start; sample < sampleCount; sample++)
            {
                buffer[sample] += buffer[sample - start] * voice.FeedbackDelay.Mix / 100;
            }
        }
    }

    private class SynthesisState
    {
        public required EnvelopeEvaluator FreqBaseEval { get; init; }
        public required EnvelopeEvaluator AmpBaseEval { get; init; }
        public EnvelopeEvaluator? FreqModRateEval { get; init; }
        public EnvelopeEvaluator? FreqModRangeEval { get; init; }
        public EnvelopeEvaluator? AmpModRateEval { get; init; }
        public EnvelopeEvaluator? AmpModRangeEval { get; init; }
        public int FrequencyStart { get; init; }
        public int FrequencyDuration { get; init; }
        public int AmplitudeStart { get; init; }
        public int AmplitudeDuration { get; init; }
        public int[] PartialDelays { get; init; } = null!;
        public int[] PartialVolumes { get; init; } = null!;
        public int[] PartialSemitones { get; init; } = null!;
        public int[] PartialStarts { get; init; } = null!;
        public EnvelopeEvaluator? FilterEnvelopeEval { get; init; }
    }
}
